### ISSUE TITLE
`AssertionError` crash on eval when val-set size is not a multiple of `train_batch_size`

### ISSUE THREAD

**Issue by jamesbraza** [NovaSky-AI/SkyRL#1583]
Since https://github.com/NovaSky-AI/SkyRL/pull/1529 in v0.2.0, SkyRL's `compute_prompt_mini_batch_boundaries` [hard-asserts](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/train/dataset/preprocess.py#L249) that the input has exactly `train_batch_size` prompts.
1. SkyRL's [`build_dataloader`](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/train/utils/trainer_utils.py#L780) sets `drop_last=False` for eval, so the eval dataloader yields a partial trailing batch whenever `len(val) % train_batch_size != 0`.
2. The training path runs with `drop_last=True`, so it never sees a partial batch
3. [The partial batch assertion](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/train/dataset/preprocess.py#L249) only fires after eval generation has already completed, and kills the whole training run

Because the `AssertionError` crash only triggers at the first eval cycle, users pay a high up-front cost (model load, possible `torch.compile`, the first `eval_interval` training steps, and the entire eval generation pass) before seeing the error. Secondly, the `assert` has no message, making the failure opaque.

```none
File ".../skyrl/train/trainer.py", line 696, in convert_to_training_input
    training_input.metadata["policy_mini_batch_boundaries"] = compute_prompt_mini_batch_boundaries(
File ".../skyrl/train/dataset/preprocess.py", line 249, in compute_prompt_mini_batch_boundaries
    assert num_prompts == train_batch_size and len(seen_uids) == train_batch_size
AssertionError
```

The crash is deterministic at the first eval cycle whenever the divisibility condition fails.

There's a few possible fixes:

- Continue to require full batches for validation, but have a validation that fails fast (before training starts)
- Relax the assertion, log a warning, and either skip the partial mini-batch or allow it

## Workaround

Upstream of SkyRL, trim the val list to `(len // train_batch_size) * train_batch_size`:

```python
kept_val, tokenized_val, _ = prepare_sft_data(val_trajectories, ..., max_length=...)
n_full = (len(kept_val) // cfg.trainer.train_batch_size) * cfg.trainer.train_batch_size
if n_full < len(kept_val):
    kept_val = kept_val[:n_full]
    tokenized_val = tokenized_val[:n_full]
```

**Comment by erictang000** [comment]
hmm @jamesbraza, can you elaborate more on how this is related to eval? My understanding is that `compute_prompt_mini_batch_boundaries` should only be called in `convert_to_training_input` during the training loop, which should only be processing batches from the training dataloader.

For example our gsm8k script seems to be running without error: https://github.com/NovaSky-AI/SkyRL/blob/main/examples/train/gsm8k/run_gsm8k.sh despite the val set size not being divisible by the eval batch size.

Could you provide some more details about your exact training script/config and I can help take a closer look. Thanks!

**Comment by jamesbraza** [comment]
You're right that stock SkyRL won't hit this, which is why the GSM8k script runs fine.

I am hitting this because I have a custom `RayPPOTrainer` subclass, with an overridden `eval` method. Within my `eval` I call `RayPPOTrainer.convert_to_training_input` and hit this `AssertionError`. I guess the resolution here can be that if I override `eval` such that I expose myself to this `AssertionError`, that I need to take care to work around it myself.

So feel free to close this issue out. That being said I still think:

1. If SkyRL can avoid having a hard `AssertionError` it would be nice, as SkyRL v0.1.0 didn't have this issue
2. Or SkyRL can populate a failure message in the `AssertionError`, to help users

**Comment by rishithayenumula** [comment]
**Root cause analysis**

The root cause is that `compute_prompt_mini_batch_boundaries` hard-asserts every batch is exactly `train_batch_size` prompts:

```python
assert num_prompts == train_batch_size and len(seen_uids) == train_batch_size
```

This holds during training (`drop_last=True`), but breaks during evaluation (`drop_last=False`), where the final batch can be smaller when `len(val) % train_batch_size != 0`. If `convert_to_training_input` is called during evaluation (e.g. in custom trainers), the partial batch triggers the `AssertionError`.

**Proposed fix**

Introduce an `is_training: bool` flag into `compute_prompt_mini_batch_boundaries` to make the behaviour explicit without duplicating logic or introducing separate code paths:

- **Training mode** — retain the strict assertion to enforce full batches
- **Eval mode** — allow partial batches, log a warning, and use `num_prompts` instead of `train_batch_size` for boundary calculations

```python
if is_training:
    assert num_prompts == train_batch_size and len(seen_uids) == train_batch_size, (
        f"Expected {train_batch_size} prompts in training batch, got {num_prompts}."
    )
else:
    if num_prompts != train_batch_size:
        logger.warning(
            f"Partial batch detected during eval: got {num_prompts} prompts, "
            f"expected {train_batch_size}. Proceeding with actual batch size."
        )
```

Additionally, all downstream loops and boundary calculations inside `compute_prompt_mini_batch_boundaries` should use `num_prompts` instead of `train_batch_size` to correctly handle partial eval batches. This prevents crashes during evaluation while preserving strict guarantees during training — minimal change, no logic duplication.

**Comment by CharlieFRuan** [comment]
> with an overridden eval method. Within my eval I call RayPPOTrainer.convert_to_training_input and hit this AssertionError

Hi @jamesbraza Could you elaborate on why you would need to convert to training input for evaluation?

Having an `is_training` flag to a `convert_to_training_input()` method seems slightly awkward to me

**Comment by rishithayenumula** [comment]
@CharlieFRuan thanks for the feedback! 

I used `convert_to_training_input()` during evaluation mainly to keep the preprocessing logic consistent between training and eval, since both paths currently expect the same input format. The `is_training` flag was intended to reuse most of the pipeline while handling minor differences.

That said, I agree this mixes training-specific assumptions into the eval path (like batch size constraints), which led to the assertion issue. It would be cleaner to separate this into a dedicated eval conversion or a shared base function without relying on `is_training`. I can refactor this accordingly.

**Comment by CharlieFRuan** [comment]
> preprocessing logic consistent between training and eval

I see, but I'm wondering why you would need `TrainingInputBatch` for evaluation (what `convert_to_training_input()` returns) instead of just `GeneratorOutput`?

**Comment by rishithayenumula** [comment]
That makes sense. As mentioned by @jamesbraza in the context of Ray operators, I reused `convert_to_training_input()` because the downstream pipeline currently expects that structure, so it was a straightforward way to keep preprocessing consistent and avoid duplicating logic.

I agree that using `TrainingInputBatch` for evaluation isn’t the right abstraction, and that evaluation should ideally operate on something like `GeneratorOutput` (or a separate eval-specific format) instead of going through training-oriented preprocessing.

Also, even if the current design assumes valid usage, it might still help to either relax the hard assertion or provide a more descriptive error message for easier debugging.

Happy to refactor this on my side to avoid routing eval through `TrainingInputBatch`.

**Comment by jamesbraza** [comment]
From my side I have a custom SFT and RL combination, and am overriding `RayPPOTrainer` methods to accomplish SFT.

I use [`TrainingInputBatch`](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/backends/skyrl_train/training_batch.py#L477) in SFT eval to measure per-trajectory cross-entropy loss, not [`GeneratorOutput`](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/train/generators/base.py#L36)'s `.rewards` (which are all-zero for SFT since there's no reward model). To compute CE I have to run a forward pass, which requires calling [`convert_to_training_input`](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/train/trainer.py#L609), as it does things like [convert prompts to tensors](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/train/trainer.py#L643-L661) and the [data-parallel padding](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/train/trainer.py#L718-L723).

I suppose in my code I can just drop my `convert_to_training_input` call and invoke `convert_prompts_responses_to_batch_tensors` and `pad_training_input_batch` directly.

I still think adding an assertion message to [`compute_prompt_mini_batch_boundaries`](https://github.com/NovaSky-AI/SkyRL/blob/skyrl-v0.2.0/skyrl/train/dataset/preprocess.py#L249) to speed debugging and guide SkyRL developers would be a nice middle-ground resolution here.

Defer to you @CharlieFRuan, and please feel free to just close this one out too.
